File: algos/greedy_algorithms/intutive_impl_A_star.py


class Node:
    def __init__(self, val = None, c=None):
        self.value = val
        self.cost = c


class Graph:
    def __init__(self):
        self.nodes = set()
        self.edges = {}
    
    def add_nodes(self, vals):
        self.nodes.update(vals)

    def add_edge(self, s, d, cost):

        if s not in self.edges:
            self.edges[s] = [Node(d, cost)]
        else:
            self.edges[s].append(Node(d, cost))

        if d not in self.edges:
            self.edges[d] = [Node(s, cost)]
        else:
            self.edges[d].append(Node(s, cost))


class MinHeap:
    def __init__(self):
        self.data = [None for _ in range(100)]
        self.next = 0

    def push(self, node):
        self.data[self.next] = node
        self.heapify_up()
        self.next +=1
        

    def pop(self):
        if not self.next :
            return None
        tmp = self.data[0]
        self.data[0] = self.data[self.next-1]
        self.data[self.next-1] = None
        self.heapify_down()
        self.next -= 1
        return tmp

    def heapify_up(self):
        cur_child = self.next
        while cur_child:
            cur_parent = (cur_child -1)//2
            if self.data[cur_child].cost < self.data[cur_parent].cost:
                tmp = self.data[cur_parent]
                self.data[cur_parent] = self.data[cur_child]
                self.data[cur_child] = tmp
                cur_child = cur_parent
            else:
                break

    def heapify_down(self):
        cur_parent = 0
        while cur_parent*2+2 < self.next-1:
            left_child = (cur_parent*2)+1
            right_child = (cur_parent*2)+2
            cur_child = left_child
            if self.data[right_child].cost < self.data[left_child].cost:
                cur_child = right_child
            if self.data[cur_child].cost < self.data[cur_parent].cost:
                tmp = self.data[cur_parent]
                self.data[cur_parent] = self.data[cur_child]
                self.data[cur_child] = tmp
                cur_parent = cur_child
            else:
                break

    def size(self):
        return self.next
        
    

graph = Graph()
graph.add_nodes(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'])
graph.add_edge('A', 'B', 2)
graph.add_edge('A', 'C', 4)
graph.add_edge('A', 'D', 1)
graph.add_edge('B', 'E', 2)
graph.add_edge('C', 'G', 1)
graph.add_edge('C', 'F', 1)
graph.add_edge('D', 'F', 1)
graph.add_edge('E', 'G', 2)
graph.add_edge('G', 'H', 3)
graph.add_edge('H', 'F', 2)

def A_star_algo(graph, source, destination):
    """Returns the smallest path from "source" to "destination"
    """
    visited = set()
    min_heap = MinHeap()
    min_heap.push(Node(source, 0))

    while min_heap.size():

        popped_node = min_heap.pop()

        if popped_node.value[-1] == destination:
            return [popped_node.value, popped_node.cost]
        
        visited.add(popped_node.value)

        if popped_node.value[-1] in graph.edges:
            neighbors = graph.edges[popped_node.value[-1]]
            for neighbor in neighbors:
                new_path = popped_node.value + neighbor.value
                new_cost = popped_node.cost + neighbor.cost
                if neighbor.value not in visited and neighbor.value not in popped_node.value:
                    min_heap.push(Node(new_path, new_cost))

    
    return None

res = A_star_algo(graph, 'A', 'H')
print(res)

res = A_star_algo(graph, 'A', 'G')
print(res)

File: algos/greedy_algorithms/intutive_impl_A_star.cpp

#include <iostream>
#include <string>
#include <queue>
#include <unordered_map>
#include <vector>
#include <set>


struct Node
{
    std::string path;
    float distance;

    Node(): Node("", 0.f){}
    Node(const std::string& p, float d): path(p), distance(d){}
};

class Graph{

    public:
    void addNodes(const std::vector<char>& vec)
    {
        for (auto node: vec)
        {
            m_nodes.insert(node);
        }
    }

    void addEdge(char s, char d, float distance)
    {
        if (m_neighbors.find(s) == m_neighbors.end())
        {
            m_neighbors[s] = std::vector<char>{d};
        }
        else
        {
            m_neighbors[s].push_back(d);
        }

        if (m_neighbors.find(d) == m_neighbors.end())
        {
            m_neighbors[d] = std::vector<char>{s};
        }
        else
        {
            m_neighbors[d].push_back(s);
        }

        std::string id1(std::string(1, s) + std::string(1, d));
        std::string id2(std::string(1, d) + std::string(1, s));
        m_distances[id1] = distance;
        m_distances[id2] = distance;
    }

    std::unordered_map<char, std::vector<char>>& getNeighbors() {return m_neighbors;}
    std::unordered_map<std::string, float>& getDistances(){return m_distances;} 

    private:
    std::unordered_map<char, std::vector<char>> m_neighbors;
    std::unordered_map<std::string, float> m_distances;
    std::set<char> m_nodes;
};

class Compare
{
    public:
    bool operator()(const Node& node1, const Node& node2)
    {
        return node1.distance > node2.distance;
    }
};

std::string AStar(Graph& graph, char s, char d)
{
    std::priority_queue<Node, std::vector<Node>, Compare> minHeap;
    minHeap.push(Node(std::string(1, s), 0.f));
    std::set<char> visited;
    std::string res("");

    while(!minHeap.empty())
    {
        auto poppedNode = minHeap.top();
        minHeap.pop();
        auto poppedId = poppedNode.path[poppedNode.path.size()-1];
        visited.insert(poppedId);
        if (poppedId == d)
        {
            res = poppedNode.path;
            break;
        }
        if (graph.getNeighbors().find(poppedId) != graph.getNeighbors().end())
        {
            auto& neighbors = graph.getNeighbors()[poppedId];
            for (auto neighbor: neighbors)
            {
                if (visited.find(neighbor) == visited.end())
                {
                    std::string disId(std::string(1, poppedId)+std::string(1, neighbor));

                    if (graph.getDistances().find(disId) != graph.getDistances().end())
                    {
                        auto dis = graph.getDistances()[disId];
                        std::string newPathName = poppedNode.path + std::string(1, neighbor);
                        float newPathDis = poppedNode.distance + dis;
                        minHeap.push(Node(newPathName, newPathDis));

                    }
                }
            }
        } 
    }
    return res;   
}



int main()
{
    Graph graph;
    std::vector<char> nodes{'A', 'B', 'C', 'D', 'M', 'S', 'D', 'L', 'F'};
    graph.addNodes(nodes);
    graph.addEdge('A', 'B', 2);
    graph.addEdge('B', 'C', 4);
    graph.addEdge('C', 'D', 6);
    graph.addEdge('A', 'M', 2);
    graph.addEdge('M', 'S', 4);
    graph.addEdge('S', 'D', 5);
    graph.addEdge('A', 'L', 2);
    graph.addEdge('L', 'M', 3);
    graph.addEdge('M', 'F', 5);
    graph.addEdge('F', 'D', 1);
   
   auto res = AStar(graph, 'A', 'D');
   std::cout<<res<<"\n";
}